turbo/transactions/call.go

```diff
@@ -124,3 +126,99 @@ func getHashGetter(requireCanonical bool, tx kv.Tx, headerReader services.Header
 		return h.Hash()
 	}
 }
+
+type ReusableCaller struct {
+	evm         *vm.EVM
+	gasCap      uint64
+	baseFee     *uint256.Int
+	stateReader state.StateReader
+}
+
+func (r *ReusableCaller) DoCall(
+	ctx context.Context,
+	args ethapi2.CallArgs,
+	callTimeout time.Duration,
+) (*core.ExecutionResult, error) {
+	var cancel context.CancelFunc
+	if callTimeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, callTimeout)
+	} else {
+		ctx, cancel = context.WithCancel(ctx)
+	}
+
+	// Make sure the context is cancelled when the call has completed
+	// this makes sure resources are cleaned up.
+	defer cancel()
+
+	msg, err := args.ToMessage(r.gasCap, r.baseFee)
+	if err != nil {
+		return nil, err
+	}
+
+	// reset the EVM so that we can continue to use it with the new context
+	txCtx := core.NewEVMTxContext(msg)
+	r.evm.Reset(txCtx, r.evm.IntraBlockState())
+
+	timedOut := false
+	go func() {
+		<-ctx.Done()
+		timedOut = true
+	}()
+
+	gp := new(core.GasPool).AddGas(msg.Gas())
+
+	result, err := core.ApplyMessage(r.evm, msg, gp, true /* refunds */, false /* gasBailout */)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the timer caused an abort, return an appropriate error message
+	if timedOut {
+		return nil, fmt.Errorf("execution aborted (timeout = %v)", callTimeout)
+	}
+
+	return result, nil
+}
+
+func NewReusableCaller(
+	stateReader state.StateReader,
+	overrides *ethapi2.StateOverrides,
+	header *types.Header,
+	initialArgs ethapi2.CallArgs,
+	gasCap uint64,
+	blockNrOrHash rpc.BlockNumberOrHash,
+	tx kv.Tx,
+	headerReader services.HeaderReader,
+	chainConfig *params.ChainConfig,
+) (*ReusableCaller, error) {
+	state := state.New(stateReader)
+
+	if overrides != nil {
+		if err := overrides.Override(state); err != nil {
+			return nil, err
+		}
+	}
+
+	var baseFee *uint256.Int
+	if header != nil && header.BaseFee != nil {
+		var overflow bool
+		baseFee, overflow = uint256.FromBig(header.BaseFee)
+		if overflow {
+			return nil, fmt.Errorf("header.BaseFee uint256 overflow")
+		}
+	}
+
+	msg, err := initialArgs.ToMessage(gasCap, baseFee)
```

Comment:
probably do `ToMessage` here and inside DoCall is too much. let's try to avoid it here.

Reply:
I could maybe expose a method on the `msg` type to adjust the gas price only as that's the only value we'll really be changing?